node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java

```diff
@@ -0,0 +1,156 @@
+    this.root = new PathPatternNode<>(IoTDBConstant.PATH_ROOT, supplier);
+    this.supplier = supplier;
+    this.appendFunction = appendFunction;
+    this.deleteFunction = deleteFunction;
+  }
+
+  /**
+   * Append key and value to PatternTreeMap.
+   *
+   * @param key PartialPath that can contain '*' or '**'
+   * @param value The value to be appended
+   */
+  public void append(PartialPath key, V value) {
+    if (appendFunction == null) {
+      throw new UnsupportedOperationException();
+    }
+    String[] pathNodes = key.getNodes();
+    PathPatternNode<V> curNode = root;
+    for (int i = 1; i < pathNodes.length; i++) {
+      PathPatternNode<V> nextNode = curNode.getChildren(pathNodes[i]);
+      if (nextNode == null) {
+        nextNode = new PathPatternNode<>(pathNodes[i], supplier);
+        curNode.addChild(nextNode);
+      }
+      curNode = nextNode;
+    }
+    curNode.appendValue(value, appendFunction);
+  }
+
+  /**
+   * Delete key and value to PatternTreeMap.
+   *
+   * @param key PartialPath that can contain '*' or '**'
+   * @param value The value to be deleted
+   */
+  public void delete(PartialPath key, V value) {
+    if (deleteFunction == null) {
+      throw new UnsupportedOperationException();
+    }
+    deletePathNode(root, key.getNodes(), 0, value);
+  }
+
+  /**
+   * Recursive method for deleting value.
+   *
+   * @param node current PathPatternNode
+   * @param pathNodes pathNodes of key
+   * @param pos current index of pathNodes
+   * @param value the value to be deleted
+   * @return true if current PathPatternNode can be removed
+   */
+  private boolean deletePathNode(PathPatternNode<V> node, String[] pathNodes, int pos, V value) {
+    if (node == null) {
+      return false;
+    }
+    if (pos == pathNodes.length - 1) {
+      node.deleteValue(value, deleteFunction);
+    } else {
+      PathPatternNode<V> child = node.getChildren(pathNodes[pos + 1]);
+      if (deletePathNode(child, pathNodes, pos + 1, value)) {
+        node.deleteChild(child);
+      }
+    }
+    return node.isLeaf() && node.getValues().isEmpty();
+  }
+
+  /**
+   * Get value list related to PathPattern that overlapped with fullPath.
+   *
+   * @param fullPath full path without wildcard
+   * @return value list
+   */
+  public List<V> getOverlapped(PartialPath fullPath) {
+    List<V> res = new ArrayList<>();
+    searchOverlapped(root, fullPath.getNodes(), 0, res, false);
+    return res;
+  }
+
+  /**
+   * Recursive method for search overlapped pattern.
+   *
+   * @param node current PathPatternNode
+   * @param pathNodes pathNodes of key
+   * @param pos current index of pathNodes
+   * @param resultList result list
+   * @param fromMultiWildCard true if node in caller is '**', so there is no need to traverse the
+   *     remaining pathNodes
+   */
+  public void searchOverlapped(
+      PathPatternNode<V> node,
+      String[] pathNodes,
+      int pos,
+      List<V> resultList,
+      boolean fromMultiWildCard) {
```

Comment:
`fromMultiWildCard` means there may exist two consecutive `**` in `PathPatternTree`?
I think this case should be deal with or avoided at the very beginning like during constructing this `PathPatternTree`.

Reply:
`fromMultiWildCard` is unnecessary. I will remove it later.